app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php

```diff
@@ -14,15 +18,36 @@ final class UpdateIntegrationUrlsRequest extends FormRequest
      */
     public function rules(): array
     {
-        $urlValidation = [
-            'id' => ['required', 'string'],
-            'url' => ['required', 'url:https', 'max:255'],
+        return [
+            'urls.*' => Rule::forEach(fn (array $value, string $attribute) => [
+                'id' => ['string'],
+                'environment' => ["required_without:$attribute.id", new Enum(Environment::class)],
+                'type' => ["required_without:$attribute.id", new Enum(IntegrationUrlType::class)],
+                'url' => ['required', 'url:https', 'max:255'],
+            ]),
         ];
+    }
 
+    public function after(): array
```

Comment:
This looks rather complex, by just looking at the code I don't know what it does 😅

Reply:
The after method is called after the main validation is done.

The error keys that will be generated are in the following structure:
`urls.0.url`

We can't map these keys to the frontend fields. Because in the frontend we group and filter the urls based on the type and environment.

What this code does, is create new error keys with the following structure:
`type.environment.url`
We also have to use `getData` to use the correct type, environment and url per url object

And after adding out custom error keys, we forget (delete) the old error keys

Reply:
In the laravel docs it says the after method should be replaced, by something else:
https://laravel.com/docs/10.x/upgrade#form-request-after-method

https://laravel.com/docs/10.x/validation#performing-additional-validation

But to me its unclear what the alternative is.

Reply:
Thanks for the extra comments. I think the first bigger comment was enough, the rest can be deduced from reading the code.